### PR TITLE
Add token in TokenSelector

### PR DESCRIPTION
**Pull request by Velenir**
Example of adding a new token through TokenSelector with minimal implementation just to see if it works
Try adding `0x45804880de22913dafe09f4980848ece6ecbaf78` on mainnet.
Not rinkeby because of id overlap

1. Open TokenSelector
2. Input token address
3. Press enter
4. Accept dialog
5. _Presto_, new token

ezgif-2-3f0bfaa8a1ee

Honestly, I don't want to deal with react-select/Creatable.

### CODE REVIEW

**Comment by gnosis-info** [comment]
Travis automatic deployment:
 https://pr780--dexreact.review.gnosisdev.com

File: src/components/TokenSelector.tsx [hunk 1]
```diff
@@ -331,17 +366,33 @@ const TokenSelector: React.FC<Props> = ({ isDisabled, tokens, selected, onChange
...
+      } else if (e.key === 'Enter' && networkId && isAddress(e.target.value.toLowerCase())) {
+        const tokenAddress = e.target.value
+        if (tokenListApi.hasToken({ tokenAddress, networkId })) return
+
+        // prevents double-catch of this event
+        // double because it's captured from onKeyDown in MenuList
+        // and in general on Select, I guess
+        e.stopPropagation()
+        console.log('tokenAddress', tokenAddress)
+        if (window.confirm(`Do you want to add ${tokenAddress} to your local tokneList`)) {
```
Comment (alfetopito) [inline src/components/TokenSelector.tsx]:
```suggestion
        if (window.confirm(`Do you want to add ${tokenAddress} to your local list of tokens?`)) {
```

File: src/components/TokenSelector.tsx [hunk 2]
```diff
@@ -331,17 +366,33 @@ const TokenSelector: React.FC<Props> = ({ isDisabled, tokens, selected, onChange
...
-    }
-  }, [])
+  const onKeyDown = useCallback(
+    (e: React.KeyboardEvent<HTMLInputElement> & { target: HTMLInputElement }): void => {
+      if (e.key === 'Escape') {
+        // Close menu on `Escape`
+        e.stopPropagation()
+        setIsFocused(false)
+        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      } else if (!e.target.value! && e.key === ' ') {
```
Comment (alfetopito) [inline src/components/TokenSelector.tsx]: Why the `!` on `.value!`?

**Comment by alfetopito** [comment]
Can't we use Modali for the confirmation here?

File: src/components/TokenSelector.tsx (hunk 2 above)
Comment (W3stside) [reply src/components/TokenSelector.tsx]: forces TS to accept it as not `falsy`

File: src/components/TokenSelector.tsx [hunk 3]
```diff
@@ -331,17 +366,33 @@ const TokenSelector: React.FC<Props> = ({ isDisabled, tokens, selected, onChange
...
+        // That closes the menu. (/shrug)
+        e.preventDefault()
+      } else if (e.key === 'Enter' && networkId && isAddress(e.target.value.toLowerCase())) {
+        const tokenAddress = e.target.value
+        if (tokenListApi.hasToken({ tokenAddress, networkId })) return
+
+        // prevents double-catch of this event
+        // double because it's captured from onKeyDown in MenuList
+        // and in general on Select, I guess
+        e.stopPropagation()
```
Comment (W3stside) [inline src/components/TokenSelector.tsx]: oof all these stopPropagations etc are getting messy. This is a `react-select` problem I'm assuming?

Comment (Velenir) [reply src/components/TokenSelector.tsx]: yes, it handles event like a blackbox, so these issues pop up from time to time

File: src/components/TokenSelector.tsx (hunk 1 above)
Comment (Velenir) [reply src/components/TokenSelector.tsx]: `window.confirm` is going away anyway

File: src/components/TokenSelector.tsx (hunk 2 above)
Comment (Velenir) [reply src/components/TokenSelector.tsx]: TS complained at some point, but not anymore, removing

File: src/components/TokenSelector.tsx [hunk 4]
```diff
@@ -298,6 +304,26 @@ const components = { MenuList }
 
 const noOptionsMessage = (): string => 'No results'
 
+const addTokenFromInput = async ({ networkId, tokenAddress }: AddTokenToListParams): Promise<boolean> => {
+  try {
+    const tokenAddedResult = await addTokenToList({ networkId, tokenAddress })
+
+    if (!isTokenAddedSuccess(tokenAddedResult)) {
+      toast.warn(tokenAddedResult.error)
```
[the remark this answers is not in the document]
Comment (anxolin) [reply src/components/TokenSelector.tsx]: Related: https://github.com/gnosis/dex-react/issues/708#issue-579806745

File: src/utils/deferred.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+export type PromiseReject = (reason?: unknown) => void
+
+export interface Deferred<T> {
+  promise: Promise<T>
+  resolve: PromiseResolve<T>
+  reject: PromiseReject
+}
+
+export const createDeferredPromise = <T>(): Deferred<T> => {
+  let resolve: PromiseResolve<T> = () => void 0
```
[the remark this answers is not in the document]
Comment (Velenir) [reply src/utils/deferred.ts:11]: `void <anything> === undefined` `void 0` is just convention, TS was complaining about empty arrow function `() => {}`

**Comment by Velenir** [comment]
> 3\. When not connected, I'm not able to add the example token. I get `token not in the exchange` message.

#812

**Comment by Velenir** [comment]
> Probably separate PR, but would it be very difficult to show a preview the token before you hit enter, as Uniswap does?

That's rather hard because the `noMessage` that is rendered there is not very controlled and supposed to be a string

**Comment by Velenir** [comment]
> 2\. As a user, I have no way to know that when I put an address in the token selector I can add it when I hit enter. Should we update the `No results` message to make that clear, or something else?

Can do, probably, would be a terrible hack though. `noOptionsMessage` is supposed to be a string. And it doesn't get when you would expect

**Comment by Velenir** [comment]
I'll merge this and other related PRs into #780 to have everything in one place
It's becoming unmanageable otherwise

**Pull request merged**